Document new templating system in Developer Guide (#4828)
ComplianceAsCode/content: docs/manual/developer_guide.adoc

is the template name.

To use a template in `rule.yml` add `template:` key there and fill it
accordingly. The general form is the following:

----
template:
    name: template_name
    vars:
        param_name: value # these parameters individual for each template
        param_name@rhel7: value1
        param_name@rhel8: value2
    backends: # optional
        ansible: off
        bash: on # on is implicit value
----

The `vars:` key contain template parameters and their values which will be
substituted into the template.  Each template has specific parameters.  To use
different values of parameters based on product, append `@` followed by product
ID to the parameter name.

The `backends` key is optional. By default, every language for which a template
with given name exist will be generated. This key can be used to explicitly opt
out from generating a certain type of content for the rule.

For example, to generate templated content except Bash remediation for rule
"Package GCC is Installed" using `package_installed` template, add the
following to `rule.yml`:

----
template:
    name: package_installed
    vars:
        pkgname: gcc
    backends:

>         bash: off

Reviewer: Unfortunately we have to quote the boolean.

Reviewer 2: That's strange, I have tested using `off` without quoting and the build system was able to switch off correctly the templated content. Do you remember what was the scenario where quoting is required?

Author: @ggbecker The problem is when you use Python 2 and you don't use quotes you will get a traceback due to some explicit "unicode" typecast in the generated processed rule.yml YAML file. 

For this moment, we will use quotes in the documentation.

But in future I wish to remove https://github.com/ComplianceAsCode/content/blob/master/ssg/yaml.py#L27 , which is causing this, but it needs to be investigated what else it could possibly break.

- `die` - Function to terminate the remediation
- `set_config_file` - Add an entry to a text configuration file

> 

Reviewer 2: I agree, you cannot find it from the table of contents, maybe increasing the table of content depth can also be beneficial for other sections.

==== Using Templates

Writing OVAL checks, Bash, or any other content can be tedious work. For
certain types of rules we provide templates. If there is a template that can be

> used for the new rule you only specify the template name and its parameters in

Reviewer 2: ```suggestion
used for the new rule you only need to specify the template name and its parameters in
```

`rule.yml` and the content will be generated during the build.

The templating system currently supports OVAL checks and Ansible, Bash,
Anaconda and Puppet remediations.  All templates can be found in
`shared/templates` directory. The files are named `template_<TYPE>_<NAME>`,
where `<TYPE>` should be OVAL, ANSIBLE, BASH, ANACONDA or PUPPET and `<NAME>`
is the template name.

To use a template in `rule.yml` add `template:` key there and fill it
accordingly. The general form is the following:

----
template:
    name: template_name
    vars:

>         param_name: value # these parameters individual for each template

Reviewer 2: Didn't understand the `these parameters individual for each template`.
Maybe:
```suggestion
        param_name: value # these parameters are individual for each template
```

        param_name@rhel7: value1
        param_name@rhel8: value2
    backends: # optional
        ansible: "off"
        bash: "on" # on is implicit value
----

> The `vars:` key contain template parameters and their values which will be

Reviewer 2: ```suggestion
The `vars:` key contains template parameters and their values which will be
```

substituted into the template.  Each template has specific parameters.  To use
different values of parameters based on product, append `@` followed by product
ID to the parameter name.

> The `backends` key is optional. By default, every language for which a template

Reviewer 2: ```suggestion
The `backends` key is optional. By default, all languages supported by a given template will be generated
```

* Languages: Ansible, Bash, OVAL

audit_rules_dac_modification::
* Checks Audit Discretionary Access Control rules
* Parameters:
** *attr* - value of `-S` argument in Audit rule, eg. `chmod`
* Languages: Ansible, Bash, OVAL

audit_rules_file_deletion_events::
* Ensure auditd Collects file deletion events
* Parameters:
** *name* - value of `-S` argument in Audit rule, eg. `unlink`
* Languages: Ansible, Bash, OVAL

audit_rules_login_events::
* Checks if there are Audit rules that record attempts to alter logon and logout events.
* Parameters:
** *path* - value of `-w` in the Audit rule, eg. `/var/run/faillock`
* Languages: Ansible, Bash, OVAL

audit_rules_path_syscall::
* Check if there are Audit rules to record events that modify user/group information via a syscall on a specific file.
* Parameters:
** *path* - path of the protected file, eg `/etc/shadow`
** *pos* - position of argument, eg. `a2`
** *syscall* - name of the system call, eg. `openat`
* Languages: Ansible, Bash, OVAL

audit_rules_privileged_commands::
* Ensure Auditd collects information on the use of privileged commands

> * Parameters:

Reviewer 3: Ensure Auditd collects information on the use of specified privileged command

** *filemode* - File permissions in a hexadecimal format, eg. `'0640'`.
* Languages: Ansible, Bash, OVAL

grub2_bootloader_argument::
* Checks kernel command line arguments in GRUB 2 configuration.
* Parameters:
** *arg_name* - argument name, eg. `audit`
** *arg_value* - argument value, eg. `'1'`
* Languages: Ansible, Bash, OVAL

kernel_module_disabled::
* Checks if the given Linux kernel module is disabled.
* Parameters:
** *kernmodule* - name of the Linux kernel module, eg. `cramfs`
* Languages: Ansible, Bash, OVAL

mount::
* Checks that a given mount point is located on a separate partition.
* Parameters:
** *mountpoint* - path to the mount point, eg. `/var/tmp`
* Languages: Anaconda, OVAL

mount_option::
* Checks if a given partition is mounted with a specific option such as "nosuid".
* Parameters:
** *mountpoint* - mount point on the filesystem eg. `/dev/shm`
** *mountoption* - mount option, eg. `nosuid`
** *filesystem* - filesystem in `/etc/fstab`, eg. `tmpfs`. Used only in Bash remediation.
** *type* - filesystem type. Used only in Bash remediation.

> ** *mount_has_to_exist* - Used only in Bash remediation. Specifies if the *mountpoint* entry has to exist in `/etc/fstab`. If set to `yes` and the *mountpoint* entry is not present in `/etc/fstab` the Bash remediation terminates.

Reviewer 3: This line confuses me. If mount has to exist (set to yes) and it is not there, I would expect the remediation to add it instead of terminating...

Author: It's really like that :smile:  I'll try to improve the wording.

* Languages: Anaconda, Ansible, Bash, OVAL

mount_option_remote_filesystems::
* Checks if all remote filesystems (NFS mounts in `/etc/fstab`) are mounted with a specific option.
* Parameters:

> ** *mountpoint* - always set to `remote_filesystems`

Reviewer 3: What is the purpose of this parameter if it is ALWAYS set to something?

Author: I don't know. I think we should take a closer look at the templates. I have opened an issue in upstream https://github.com/ComplianceAsCode/content/issues/4914

** *mountoption* - mount option, eg. `nodev`
** *filesystem* - filesystem of new mount point (used when adding new entry in `/etc/fstab`), eg. `tmpfs`. Used only in Bash remediation.
* Languages: Ansible, Bash, OVAL

mount_option_removable_partitions::
* Checks if all removable media mounts are mounted with a specific option.
* Parameters:

> ** *mountpoint* - always set to `remote_filesystems`

Reviewer 3: What is the purpose of this parameter? Why talking about remote_systems when the template is concerning removable filesystems?

Author: this is actually a copy-paste error, it should be always set to `var_removable_partition`

** *mountoption* - mount option, eg. `nodev`
** *filesystem* - filesystem of new mount point (used when adding new entry in `/etc/fstab`), eg. `tmpfs`. Used only in Bash remediation.
* Languages: Anaconda, Ansible, Bash, OVAL

package_installed::

> * Checks if a given package is installed. Optionally, it can also check whether a specific version or newer is installed.

Reviewer 3: In the parameters below, the evr parameter is said to check for specific value. How to specify that we want newer version, eg greater or equal to the evr?

Author: Actually, it checks if the given version or a newer version is installed, because OVAL check uses  operation "greater than or equal".    It's here:  https://github.com/ComplianceAsCode/content/blob/85bdd8bbb8753f77f6c728811c89f026b76b4d8b/shared/macros-oval.jinja#L371

At this moment it isn't possible to parametrize the operation, eg. to check for older version or for the exactly the same version.

I'll try to improve the wording.

* Parameters:
** *pkgname* - name of the RPM or DEB package, eg. `tmux`

> ** *evr* - Optional parameter. Use to check a specific epoch, version, release in `epoch:version-release` format. Used only in OVAL checks.

Reviewer 3: Could you please give example of such parameter?

Author: yes

* Languages: Anaconda, Ansible, Bash, OVAL, Puppet

package_removed::
* Checks if the given package is not installed.
* Parameters:
** *pkgname* - name of the RPM or DEB package, eg. `tmux`
* Languages: Anaconda, Ansible, Bash, OVAL, Puppet

sebool::
* Checks values of SELinux booleans.
* Parameters:
** *seboolid* - name of SELinux boolean, eg. `cron_userdomain_transition`

> ** *sebool_bool* - the value of the SELinux Boolean. Can be either `"true"` or `"false"`. If this parameter is not specified, the rule will use XCCDF Value `var_<seboolid>`

Reviewer 3: Could you explain more about var_seboolid? Where is it specified, what is it? Or is it supposed that an user already knows that?

Reviewer 2: This is actually the usual XCCDF variables that you can define in profiles, there might be some documentation available on how to use them but I'm not sure where to right now. The seboolid will be replaced by some selinux boolean, for example: selinuxuser_execheap and in the profile you can use var_selinuxuser_execheap to turn on or off the boolean.

Reviewer 2: @jan-cerny Referencing the section which tells on how to use variables can be a good thing here.

Author: Do we have any section like that?

Reviewer 2: To be honest I don't know, I just assumed there was something.

* Languages: Ansible, Bash, OVAL, Puppet

service_enabled::
* Checks if a system service is enabled. Uses either systemd or SysV init based on the product configuration in `product.yml`.
* Parameters:
** *servicename* - name of the service.
** *packagename* - name of the package that provides this service. This argument is optional. If *packagename* is not specified it means the name of the package is the same as the name of service.
** *daemonname* - name of the daemon. This argument is optional. If *daemonname* is not specified it means the name of the daemon is the same as the name of service.
* Languages: Ansible, Bash, OVAL, Puppet

sshd_lineinfile::
* Checks SSH server configuration items in `/etc/ssh/sshd_config`.
* Parameters:
** *parameter* - name of the SSH configuration option, eg. `KerberosAuthentication`
** *value* - value of the SSH configuration option specified by *parameter*, eg. `"no"`.
** *missing_parameter_pass* - If set to `"true"` the OVAL check will pass if the parameter is not present in `/etc/ssh/sshd_config`.
* Languages: Ansible, Bash, OVAL

sysctl::
* Checks sysctl parameters.

> * Parameters:

Reviewer 3: Is it worth mentioning where are these parameters checked? file names, etc?

** *sysctlvar* - name of the sysctl value, eg. `net.ipv4.conf.all.secure_redirects`.
** *datatype* - data type of the sysctl value, eg. `int`.
** *sysctlval* - value of the sysctl value, eg. `'1'`. If this parameter is not specified, XCCDF Value is used instead.
* Languages: Ansible, Bash, OVAL

timer_enabled::
* Checks if a SystemD timer unit is enabled.
* Parameters:
** *timername* - name of the SystemD timer unit, without the `timer` suffix, eg. `dnf-automatic`.

> ** *packagename* - name of the RPM package which provides the SystemD timer unit. This parameter is optional, if it is not provideded it is assumed that the name of the RPM package is the same as the name of the SystemD timer unit.

Reviewer 3: just a simple mistake, not provideded

* Languages: Ansible, Bash, OVAL

==== Creating Templates

The offer of currently available templates can be extended by developing a new
template.

1) Create the template files, one for each type of file. Each one should be
named `template_<TYPE>_<NAME>`. Where `<TYPE>` should be OVAL, ANSIBLE, BASH,
ANACONDA or PUPPET and `<NAME>` is the what we will call the template name.

Use the Jinja syntax we use elsewhere in the project; refer to the earlier
section on Jinja macros for more information.  The parameters should be named
using lowercase letters.

Notice that OVAL should be written in shorthand format.  This is an example of
an OVAL template file called _template_OVAL_package_installed_:

> 

Reviewer 3: I am confused. Why does the file name has underscores at the begining and end of name? Is this something adoc specific?

Reviewer 2: It is, it formats the text to italic.

----

> include::{templatesdir}/template_OVAL_package_installed

Reviewer 3: Is this supposed to be verbatim? Or should it show the content of some file? If yes the content is not shown, at least not on github.

Reviewer 2: From adoc docs: https://github.com/asciidoctor/asciidoctor.org/blob/master/docs/_includes/include-directive.adoc
```suggestion
include::{templatesdir}/template_OVAL_package_installed[]
```

Reviewer 2: @jan-cerny please fix all `include::` directives

----

And here is the Ansible template file called _template_ANSIBLE_package_installed_:

----
include::{templatesdir}/template_ANSIBLE_package_installed
----

2) Implement a callback function which will process the
parameters before passing them to Jinja engine.  For example, this callback can
provide default values, escape characters, check if parameters are correct, or
any other processing of the parameters specific for the template.  The callback

> function should be put into `ssg/templates.py` and must have 2 parameters:

Reviewer 2: ```suggestion
function should be put into `link:{rootdir}/ssg/templates.py[templates.py]` and must have 2 parameters:
```

- `data` - will contain the contents of `vars:` dictionary from `rule.yml`

> - `lang` - the language (`anaconda`, `ansible`, `bash`, `oval`, `puppet`)

Reviewer 3: How is this parameter used? Will it be populated with some list / array?

Reviewer 2: @jan-cerny does the callback function get executed once for every supported language for current template? So in theory I could process the data differently for each supported language. Is that the current behavior?

Author: @vojtapolasek it's a string, it can be exactly one of the aforementioned values, the function is called for each supported language.

@ggbecker Yes, it gets executed once for every supported language for current template. Yes, you can process data differently, that is the current behavior, it's done like that in some of the callbacks.

Author: I'll try to improve the wording

The function must always return the (modified) `data` dictionary.

The function must be always defined even if no processing of data is needed.
In that situation the function just returns `data` parameter.

3) Add the template name to `templates` dictionary in `ssg/templates.py`. This

> dictionary maps template name to the callback function name.

Reviewer 2: ```suggestion
dictionary maps the template name to the callback function name.
```

* Parameters:
** *path* - the path of the privileged command - eg. `/usr/bin/mount`
* Languages: Ansible, Bash, OVAL

audit_rules_unsuccessful_file_modification::
* Ensure there is an Audit rule to record unsuccessful access attempts to files
* Parameters:
** *name* - name of the unsuccessful system call, eg. `creat`
* Languages: Ansible, Bash, OVAL

audit_rules_unsuccessful_file_modification_o_creat::
* Ensure there is an Audit rule to record unsuccessful access attempts to files when O_CREAT flag is specified.
* Parameters:
** *syscall* - name of the unsuccessful system call, eg. `openat`
** *pos* - position of the O_CREAT argument in the syscall, as specified by `-F` audit rule argument, eg. `a2`
* Languages: OVAL

audit_rules_unsuccessful_file_modification_o_trunc_write::
* Ensure there is an Audit rule to record unsuccessful access attempts to files when O_TRUNC_WRITE flag is specified.
* Parameters:
** *syscall* - name of the unsuccessful system call, eg. `openat`
** *pos* - position of the O_TRUNC_WRITE argument in the syscall, as specified by `-F` audit rule argument, eg. `a2`
* Languages: OVAL

audit_rules_unsuccessful_file_modification_rule_order::

> * Ensure that Audit rules for unauthorized attempts to a specific system call are ordered correctly.

Reviewer 3: ```suggestion
* Ensure that Audit rules for unauthorized attempts to use a specific system call are ordered correctly.
```

* Languages: Anaconda, Ansible, Bash, OVAL

mount_option_remote_filesystems::
* Checks if all remote filesystems (NFS mounts in `/etc/fstab`) are mounted with a specific option.
* Parameters:
** *mountpoint* - always set to `remote_filesystems`
** *mountoption* - mount option, eg. `nodev`
** *filesystem* - filesystem of new mount point (used when adding new entry in `/etc/fstab`), eg. `tmpfs`. Used only in Bash remediation.
** *mount_has_to_exist* - Used only in Bash remediation. Specifies if the *mountpoint* entry has to exist in `/etc/fstab` before the remediation is executed. If set to `yes` and the *mountpoint* entry is not present in `/etc/fstab` the Bash remediation terminates. If set to `no` the *mountpoint* entry will be created in `/etc/fstab`.
* Languages: Ansible, Bash, OVAL

mount_option_removable_partitions::
* Checks if all removable media mounts are mounted with a specific option.
* Parameters:

> ** *mountpoint* - always set to `var_removable_partition`

Reviewer 3: is this an XCCDF variable?

Author: yes, I'll add it

** *packagename* - name of the package that provides this service. This argument is optional. If *packagename* is not specified it means the name of the package is the same as the name of service.
** *daemonname* - name of the daemon. This argument is optional. If *daemonname* is not specified it means the name of the daemon is the same as the name of service.
* Languages: Ansible, Bash, OVAL, Puppet

sshd_lineinfile::
* Checks SSH server configuration items in `/etc/ssh/sshd_config`.
* Parameters:
** *parameter* - name of the SSH configuration option, eg. `KerberosAuthentication`
** *value* - value of the SSH configuration option specified by *parameter*, eg. `"no"`.
** *missing_parameter_pass* - If set to `"true"` the OVAL check will pass if the parameter is not present in `/etc/ssh/sshd_config`.
* Languages: Ansible, Bash, OVAL

sysctl::
* Checks sysctl parameters. The OVAL definition checks both configuration and runtime settings and require both of them to be set to the desired value to return true.
* Parameters:
** *sysctlvar* - name of the sysctl value, eg. `net.ipv4.conf.all.secure_redirects`.
** *datatype* - data type of the sysctl value, eg. `int`.

> ** *sysctlval* - value of the sysctl value, eg. `'1'`. If this parameter is not specified, XCCDF Value is used instead.

Reviewer 3: Do you mean xccdf variable?

Author: Yes, I mean a variable, but unfortunately, variables in XCCDF terminology are called XCCDF Values, and they're represented by `<xccdf:Value>` element :scream:

* Languages: Ansible, Bash, OVAL

timer_enabled::
* Checks if a SystemD timer unit is enabled.
* Parameters:
** *timername* - name of the SystemD timer unit, without the `timer` suffix, eg. `dnf-automatic`.
** *packagename* - name of the RPM package which provides the SystemD timer unit. This parameter is optional, if it is not provided it is assumed that the name of the RPM package is the same as the name of the SystemD timer unit.
* Languages: Ansible, Bash, OVAL

==== Creating Templates

The offer of currently available templates can be extended by developing a new
template.

1) Create the template files, one for each type of file. Each one should be
named `template_<TYPE>_<NAME>`. Where `<TYPE>` should be OVAL, ANSIBLE, BASH,
ANACONDA or PUPPET and `<NAME>` is the what we will call the template name.
Create these files in `link:{rootdir}/shared/templates[shared/templates]` directory.

Use the Jinja syntax we use elsewhere in the project; refer to the earlier
section on Jinja macros for more information.  The parameters should be named

> using lowercase letters.

Reviewer 3: What do you mean by named using lower case letters? I am confused because in the example oval template the parameters are in upper case.

Author: Should be uppercase there at this moment. I'll fix that. But I think using lowercase letters this is the desired future state, because using uppercase letters is only a temporary workaround for easier transition from old system to the new system.

https://github.com/ComplianceAsCode/content/blob/85bdd8bbb8753f77f6c728811c89f026b76b4d8b/ssg/templates.py#L318-L323

----
# platform = multi_platform_all
# reboot = false
# strategy = enable
# complexity = low
# disruption = low
- name: Ensure {{{ PKGNAME }}} is installed
  package:
    name: "{{{ PKGNAME }}}"
    state: present

----

2) Implement a callback function which will process the
parameters before passing them to the Jinja engine.  For example, this callback can
provide default values, escape characters, check if parameters are correct, or
any other processing of the parameters specific for the template.

The callback function should be put into `link:{rootdir}/ssg/templates.py[ssg/templates.py]` and must have 2 parameters:

- `data` - dictionary which contains the contents of `vars:` dictionary from `rule.yml`
- `lang` - string, describes language, can be one of: `"anaconda"`, `"ansible"`, `"bash"`, `"oval"`, `"puppet"`

The callback function is executed for every supported language, so it can process the data differently for each language.

The function must always return the (modified) `data` dictionary.

The function must be always defined even if no processing of data is needed.
In that situation the function just returns `data` parameter.

> The following example shows callback for template `mount_option`:

Reviewer 2: ```suggestion
The following example shows the callback function for template `mount_option`:
```
